torch_xla/experimental/xla_sharded_tensor.py

```diff
@@ -109,12 +109,9 @@ def __new__(cls, elem: torch.Tensor, *args, **kwargs):
   # which results from the sharding.
   @property
   def local_shards(self) -> List[XLAShard]:
-    shards = torch_xla._XLAC._get_local_shards(self.global_tensor)
-    devices = [str(shard.device) for shard in shards]
+    shards, devices = torch_xla._XLAC._get_local_shards(self.global_tensor)
     indices = torch_xla._XLAC._get_local_shard_indices(self.global_tensor)
-    return [
-        XLAShard(s.cpu(), i, d) for s, i, d in zip(shards, indices, devices)
-    ]
+    return [XLAShard(s, i, d) for s, i, d in zip(shards, indices, devices)]
```

Comment:
It looks like `XLAShard.shard_device` will be used by `torch_xla._XLAC._load_local_shards`. Then within, it calls `GetXlaDevices` which uses `bridge::AtenDeviceToXlaDevice(c10::Device(device_str))` to create the backend data. I wonder how this works now after you return tpu:0 instead of xla:0.

Reply:
we used to convert `tpu:0` -> `xla:0` and return `xla:0` as `XLAShards`'s device. In the load part we call `GetXlaDevices` to convert `xla:0` back to `tpu:0`.

After this change we directly save `tpu:0` as the `XLAShard`'s device. The downside is if we saved a shard on TPU, we can't load it back to GPU since now device is hard code as TPU. This is fixable through, I just don't want to expose `xla:1` -> `xla:3` to user since there should only one `xla:0` which is a virtual device.

Reply:
By reading the code, I believe `device_str` in `c10::Device(device_str)` is `tpu:0`? Then how can `c10::Device` parse it?

Reply:
Just a note - we don't serialize the device string (e.g. `tpu:0`) when checkpointing, the device string is only tracked so we can map the CPU shard back to the shard on device. This change doesn't break anything for checkpointing.